// forest-express/dist/services/permissions-formatter.js
"use strict";

var _transformActionsPermissionsFromOldToNewFormat = function _transformActionsPermissionsFromOldToNewFormat(smartActionsPermissions) {
  var newSmartActionsPermissions = {};
  Object.keys(smartActionsPermissions).forEach(function (actionName) {
    var action = smartActionsPermissions[actionName];
    newSmartActionsPermissions[actionName] = {
      triggerEnabled: action.users ? action.allowed && action.users : action.allowed
    };
  });
  return newSmartActionsPermissions;
};

var transformPermissionsFromOldToNewFormat = function transformPermissionsFromOldToNewFormat(permissions) {
  var newPermissions = {};
  Object.keys(permissions).forEach(function (modelName) {
    var modelPermissions = permissions[modelName];
    var collection = modelPermissions.collection;
    newPermissions[modelName] = {
      collection: {
        browseEnabled: collection.list || collection.searchToEdit,
        readEnabled: collection.show,
        editEnabled: collection.update,
        addEnabled: collection.create,
        deleteEnabled: collection["delete"],
        exportEnabled: collection["export"]
      },
      scope: modelPermissions.scope
    };

    if (modelPermissions.actions) {
      newPermissions[modelName].actions = _transformActionsPermissionsFromOldToNewFormat(modelPermissions.actions);
    }
  });
  return newPermissions;
};

module.exports = {
  transformPermissionsFromOldToNewFormat: transformPermissionsFromOldToNewFormat
};
