// forest-express/dist/context/init.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var fs = require('fs');

var moment = require('moment');

var VError = require('verror');

var superagentRequest = require('superagent');

var path = require('path');

var openIdClient = require('openid-client');

var jsonwebtoken = require('jsonwebtoken');

var errorMessages = require('../utils/error-messages');

var errorUtils = require('../utils/error');

var stringUtils = require('../utils/string');

var isSameDataStructure = require('../utils/is-same-data-structure');

var joinUrl = require('../utils/join-url');

var _require = require('../utils/widgets'),
    setFieldWidget = _require.setFieldWidget;

var logger = require('../services/logger');

var pathService = require('../services/path');

var errorHandler = require('../services/exposed/error-handler');

var ipWhitelist = require('../services/ip-whitelist');

var forestServerRequester = require('../services/forest-server-requester');

var ApimapSorter = require('../services/apimap-sorter');

var ApimapSender = require('../services/apimap-sender');

var ApimapFieldsFormater = require('../services/apimap-fields-formater');

var AuthorizationFinder = require('../services/authorization-finder');

var baseFilterParser = require('../services/base-filters-parser');

var ConfigStore = require('../services/config-store');

var PermissionsChecker = require('../services/permissions-checker');

var PermissionsGetter = require('../services/permissions-getter');

var permissionsFormatter = require('../services/permissions-formatter');

var SchemaFileUpdater = require('../services/schema-file-updater');

var SmartActionHook = require('../services/smart-action-hook');

var schemasGenerator = require('../generators/schemas');

var ModelsManager = require('../services/models-manager');

var AuthenticationService = require('../services/authentication');

var TokenService = require('../services/token');

var OidcConfigurationRetrieverService = require('../services/oidc-configuration-retriever');

var OidcClientManagerService = require('../services/oidc-client-manager');

var ProjectDirectoryFinder = require('../services/project-directory-finder');

function initValue(context) {
  context.addValue('forestUrl', process.env.FOREST_URL || 'https://api.forestadmin.com');
}
/**
 * @typedef {{
 *   NODE_ENV: 'production' | 'development';
 *   FOREST_DISABLE_AUTO_SCHEMA_APPLY: boolean;
 *   CORS_ORIGINS?: string;
 *   JWT_ALGORITHM: string;
 *   FOREST_PERMISSIONS_EXPIRATION_IN_SECONDS: number;
 *   FOREST_OIDC_CONFIG_EXPIRATION_IN_SECONDS: number;
 *   FOREST_URL: string;
 *   APPLICATION_URL: string;
 *   FOREST_AUTH_SECRET: string;
 *   FOREST_ENV_SECRET: string;
 * }} Env
 *
 * @typedef {{
 *  env: Env
 * }} EnvPart
 *
 * @typedef {{
 *  errorMessages: import('../utils/error-messages');
 *  stringUtils: import('../utils/string');
 *  errorUtils: import('../utils/error');
 *  isSameDataStructure: import('../utils/object-have-same-keys')
 *  setFieldWidget: import('../utils/widgets').setFieldWidget
 *  joinUrl: import('../utils/join-url')
 * }} Utils
 *
 * @typedef {{
 *  logger: import('../services/logger');
 *  pathService: import('../services/path');
 *  errorHandler: import('../services/exposed/error-handler');
 *  ipWhitelist: import('../services/ip-whitelist');
 *  forestServerRequester: import('../services/forest-server-requester');
 *  authorizationFinder: import('../services/authorization-finder');
 *  schemaFileUpdater: import('../services/schema-file-updater');
 *  apimapSender: import('../services/apimap-sender');
 *  permissionsChecker: import('../services/permissions-checker');
 *  permissionsGetter: import('../services/permissions-getter');
 *  smartActionHook: import('../services/smart-action-hook');
 *  schemasGenerator: import('../generators/schemas');
 *  authenticationService: import('../services/authentication');
 *  tokenService: import('../services/token');
 *  oidcConfigurationRetrieverService: import('../services/oidc-configuration-retriever');
 *  oidcClientManagerService: import('../services/oidc-client-manager')
 *  configStore: import('../services/config-store')
 * }} Services
 *
 * @typedef {{
 *  superagentRequest: import('superagent');
 *  openIdClient: import('openid-client');
 *  jsonwebtoken: import('jsonwebtoken');
 * }} Externals
 *
 * @typedef {EnvPart & Utils & Services & Externals} Context
 */

/**
 * @param {import('./application-context')} context
 */


function initEnv(context) {
  context.addInstance('env', _objectSpread(_objectSpread({}, process.env), {}, {
    FOREST_URL: process.env.FOREST_URL || 'https://api.forestadmin.com',
    JWT_ALGORITHM: process.env.JWT_ALGORITHM || 'HS256',
    NODE_ENV: ['dev', 'development'].includes(process.env.NODE_ENV) ? 'development' : 'production',
    APPLICATION_URL: process.env.APPLICATION_URL || "http://localhost:".concat(process.env.APPLICATION_PORT || 3310)
  }));
}
/**
 * @param {import('./application-context')} context
 */


function initUtils(context) {
  context.addInstance('errorMessages', errorMessages);
  context.addInstance('stringUtils', stringUtils);
  context.addInstance('errorUtils', errorUtils);
  context.addInstance('isSameDataStructure', isSameDataStructure);
  context.addInstance('setFieldWidget', setFieldWidget);
  context.addInstance('joinUrl', joinUrl);
}
/**
 * @param {ApplicationContext} context
 */


function initServices(context) {
  context.addInstance('logger', logger);
  context.addInstance('pathService', pathService);
  context.addInstance('errorHandler', errorHandler);
  context.addInstance('ipWhitelist', ipWhitelist);
  context.addInstance('forestServerRequester', forestServerRequester);
  context.addInstance('schemasGenerator', schemasGenerator);
  context.addInstance('baseFilterParser', baseFilterParser);
  context.addInstance('permissionsFormatter', permissionsFormatter);
  context.addClass(ProjectDirectoryFinder);
  context.addClass(ConfigStore);
  context.addClass(PermissionsGetter);
  context.addClass(PermissionsChecker);
  context.addClass(ApimapFieldsFormater);
  context.addClass(AuthorizationFinder);
  context.addClass(ApimapSorter);
  context.addClass(ApimapSender);
  context.addClass(SchemaFileUpdater);
  context.addClass(ModelsManager);
  context.addClass(TokenService);
  context.addClass(OidcConfigurationRetrieverService);
  context.addClass(OidcClientManagerService);
  context.addClass(AuthenticationService);
  context.addClass(SmartActionHook);
}
/**
 * @param {import('./application-context')} context
 */


function initExternals(context) {
  context.addInstance('superagentRequest', superagentRequest);
  context.addInstance('fs', fs);
  context.addInstance('path', path);
  context.addInstance('openIdClient', openIdClient);
  context.addInstance('jsonwebtoken', jsonwebtoken);
  context.addInstance('moment', moment);
  context.addInstance('VError', VError);
}
/**
 * @param {import('./application-context')<Context>} context
 */


function initContext(context) {
  initExternals(context);
  initValue(context);
  initEnv(context);
  initUtils(context);
  initServices(context);
}

module.exports = initContext;
